drivers/audio/dmic_nrfx_pdm.c

```diff
@@ -65,17 +65,25 @@ static void event_handler(const struct device *dev, const nrfx_pdm_evt_t *evt)
 	const struct dmic_nrfx_pdm_drv_cfg *drv_cfg = dev->config;
 	int ret;
 	bool stop = false;
+	void *mem_slab_buffer;
 
 	if (evt->buffer_requested) {
 		void *buffer;
 		nrfx_err_t err;
 
-		ret = k_mem_slab_alloc(drv_data->mem_slab, &drv_data->mem_slab_buffer, K_NO_WAIT);
+		ret = k_mem_slab_alloc(drv_data->mem_slab, &mem_slab_buffer, K_NO_WAIT);
 		if (ret < 0) {
 			LOG_ERR("Failed to allocate buffer: %d", ret);
 			stop = true;
 		} else {
-			ret = dmm_buffer_in_prepare(drv_cfg->mem_reg, drv_data->mem_slab_buffer,
+			ret = k_msgq_put(&drv_data->mem_slab_queue, &mem_slab_buffer, K_NO_WAIT);
+			if (ret < 0) {
+				LOG_ERR("No room in mem slab queue");
+				free_buffer(drv_data, mem_slab_buffer);
+				stop_pdm(drv_data);
+				return;
+			}
+			ret = dmm_buffer_in_prepare(drv_cfg->mem_reg, mem_slab_buffer,
 						    drv_data->block_size, &buffer);
 			if (ret < 0) {
 				LOG_ERR("Failed to prepare buffer: %d", ret);
```

Comment:
aren't we missing `free_buffer` here? additionally if `dmm_buffer_in_prepare` fails then `mem_slab_queue` contains buffer that will never be processed.

Shouldn't we first check if there is a room in `mem_slab_queue`, then call `dmm_buffer_in_prepare`, then call `k_msgq_put` and assert on error code (this assertion would mean we have another producer on `mem_slab_queue`, which is incorrect) ?